propolis/src/hw/nvme/bits.rs

```diff
@@ -0,0 +1,365 @@
+pub const STS_FAILED_FUSED: u8 = 0x9;
+pub const STS_MISSING_FUSED: u8 = 0xa;
+pub const STS_INVALID_NS: u8 = 0xb;
+pub const STS_COMMAND_SEQ_ERR: u8 = 0xc;
+pub const STS_INVAL_SGL_DESC: u8 = 0xd;
+pub const STS_INVAL_SGL_NUM: u8 = 0xe;
+pub const STS_INVAL_SGL_LEN: u8 = 0xf;
+pub const STS_INVAL_SGL_META_LEN: u8 = 0x10;
+pub const STS_INVAL_SGL_TYPE_LEN: u8 = 0x11;
+pub const STS_INVAL_CMB_USE: u8 = 0x12;
+pub const STS_INVAL_PRP_OFFSET: u8 = 0x13;
+
+pub const STS_CREATE_IO_Q_INVAL_CQ: u8 = 0x0;
+pub const STS_CREATE_IO_Q_INVAL_QID: u8 = 0x1;
+pub const STS_CREATE_IO_Q_INVAL_QSIZE: u8 = 0x2;
+pub const STS_CREATE_IO_Q_INVAL_INT_VEC: u8 = 0x8;
+
+// SetFeature Command Specific Status values
+
+pub const STS_SET_FEATURE_NOT_SAVEABLE: u8 = 0x0D;
+pub const STS_SET_FEATURE_NOT_CHANGEABLE: u8 = 0x0E;
+pub const STS_SET_FEATURE_NOT_NAMESPACE_SPECIFIC: u8 = 0x0F;
+pub const STS_SET_FEATURE_OVERLAPPING_RANGES: u8 = 0x14;
+
+// Read Command Status values
+
+pub const STS_READ_CONFLICTING_ATTRS: u8 = 0x80;
+pub const STS_READ_INVALID_PROT_INFO: u8 = 0x81;
+
+// Feature identifiers
+
+pub const FEAT_ID_ARBITRATION: u8 = 0x01;
+pub const FEAT_ID_POWER_MGMT: u8 = 0x02;
+pub const FEAT_ID_TEMP_THRESH: u8 = 0x04;
+pub const FEAT_ID_ERROR_RECOVERY: u8 = 0x05;
+pub const FEAT_ID_NUM_QUEUES: u8 = 0x07;
+pub const FEAT_ID_INTR_COALESCE: u8 = 0x08;
+pub const FEAT_ID_INTR_VEC_CFG: u8 = 0x09;
+pub const FEAT_ID_WRITE_ATOMIC: u8 = 0x0a;
+pub const FEAT_ID_ASYNC_EVENT_CFG: u8 = 0x0b;
+
+// Identify CNS values
+
+pub const IDENT_CNS_NAMESPACE: u8 = 0x0;
+pub const IDENT_CNS_CONTROLLER: u8 = 0x1;
+pub const IDENT_CNS_ACTIVE_NSID: u8 = 0x2;
+pub const IDENT_CNS_NSID_DESC: u8 = 0x3;
+
+#[derive(Copy, Clone)]
+#[repr(u8)]
+pub enum StatusCodeType {
+    Generic = 0,
+    CmdSpecific = 1,
+    MediaDataIntegrity = 2,
+    VendorSpecific = 7,
+}
+
+#[derive(Default, Copy, Clone)]
+#[repr(C)]
+pub struct PowerStateDescriptor {
+    /// Maximum Power
+    pub mp: u16,
+    /// Reserved
+    pub _resv1: u16,
+    /// Entry Latency
+    pub enlat: u32,
+    /// Exit Latency
+    pub exlat: u32,
+    /// Relative Read Throughput
+    pub rrt: u8,
+    /// Relative Read Latency
+    pub rrl: u8,
+    /// Relative Write Throughput
+    pub rwt: u8,
+    /// Relative Write Latency
+    pub rwl: u8,
+    /// Reserved
+    pub _resv: [u8; 16],
+}
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct IdentifyController {
+    // bytes 0-255 - Controller Capabilities and Features
+    /// PCI Vendor ID
+    pub vid: u16,
+    /// PCI Subsystem Vendor ID
+    pub ssvid: u16,
+    /// Serial Number
+    pub sn: [u8; 20],
+    /// Model Number
+    pub mn: [u8; 40],
+    /// Firmware Revision
+    pub fr: [u8; 8],
+    /// Recommended Arbitration Burst
+    pub rab: u8,
+    /// IEEE OUI Identifier
+    pub ieee: [u8; 3],
+    /// Multi-Interface Capabilities
+    pub cmic: u8,
+    /// Maximum Data Transfer Size
+    pub mdts: u8,
+    /// Reserved
+    pub _resv1: [u8; 178],
```

Comment:
One minor suggestion from maintaining these in C, if it's possible to phrase it like the spec, e.g. 256-78, I found that it also led to fewer errors being committed by folks when they were updating these. That way someone could see the field in the spec and go from there.